File: Problem_1.py
"""
Problem : 1

Time Complexity : 
Approach 1 - O(m*n)
Approach 2 - O(m*n)

Space Complexity : 
Approach 1 - O(m*n)
Approach 2 - O(m*n)

Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No


"""

# Minesweeper

# Approach - 1
# DFS

class Solution(object):
    def __init__(self):
        self.directions=[(1,0),(1,1),(0,1),(-1,1),(-1,0),(-1,-1),(0,-1),(1,-1)]
    def updateBoard(self, board, click):
        """
        :type board: List[List[str]]
        :type click: List[int]
        :rtype: List[List[str]]
        """
        if not board:
            return board
        m=len(board)
        n=len(board[0])

        if board[click[0]][click[1]]=='M':
            board[click[0]][click[1]]='X'
            return board

        self.dfs(board,click[0],click[1],m,n)

        return board
    
    def dfs(self,board,i,j,m,n):
        # base
        if i<0 or i>=m or j<0 or j>=n or board[i][j]!='E':
            return 
        # logic
        count=self.countMines(board,i,j,m,n)
        if count==0:
            board[i][j]='B'
            for dirs in self.directions:
                nr=i+dirs[0]
                nc=j+dirs[1]
                self.dfs(board,nr,nc,m,n)
        else:
            board[i][j]=str(count)

    def countMines(self,board,i,j,m,n):
        count=0
        for dirs in self.directions:
            nr=i+dirs[0]
            nc=j+dirs[1]
            if nr>=0 and nr<m and nc>=0 and nc<n and board[nr][nc]=='M':
                    count+=1
        return count


# Approach - 2
# BFS

class Solution(object):
    def __init__(self):
        self.directions=[(1,0),(1,1),(0,1),(-1,1),(-1,0),(-1,-1),(0,-1),(1,-1)]
    def updateBoard(self, board, click):
        """
        :type board: List[List[str]]
        :type click: List[int]
        :rtype: List[List[str]]
        """
        if not board:
            return board
        m=len(board)
        n=len(board[0])

        if board[click[0]][click[1]]=='M':
            board[click[0]][click[1]]='X'
            return board

        q=collections.deque()
        q.append((click[0],click[1]))
        board[click[0]][click[1]]='B'

        while q:
            curr=q.popleft()
            count=self.countMines(board,curr[0],curr[1],m,n)
            if count==0:
                for dirs in self.directions:
                    nr=curr[0]+dirs[0]
                    nc=curr[1]+dirs[1]
                    if nr>=0 and nr<m and nc>=0 and nc<n and board[nr][nc]=='E':
                        q.append((nr,nc))
                        board[nr][nc]='B'
            else:
                board[curr[0]][curr[1]]=str(count)
        return board
    def countMines(self,board,i,j,m,n):
        count=0
        for dirs in self.directions:
            nr=i+dirs[0]
            nc=j+dirs[1]
            if nr>=0 and nr<m and nc>=0 and nc<n and board[nr][nc]=='M':
                    count+=1
        return count

File: Problem_1.js
// Problem1: Minesweeper (https://leetcode.com/problems/minesweeper/)

// Time Complexity : O(V+E)
// Space Complexity : O(V)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

let dirs, m, n;
var getMines = (board, r, c) => {
    let count = 0;
    for (let i = 0; i < dirs.length; i++) {
        let nr = r + dirs[i][0];
        let nc = c + dirs[i][1];
        if (nr >= 0 && nc >= 0 && nr < m && nc < n && board[nr][nc] === 'M')
            count++;
    }
    return count;
}

var dfs = (board, r, c) => {
    // Base case
    if (r < 0 || c < 0 || r >= m || c >= n || board[r][c] !== 'E')
        return;
    // Logic
    let mineCount = getMines(board, r, c);
    if (mineCount > 0) {
        board[r][c] = mineCount.toString();
    } else {
        board[r][c] = "B";
        for (let i = 0; i < dirs.length; i++) {
            let nr = r + dirs[i][0];
            let nc = c + dirs[i][1];
            dfs(board, nr, nc);
        }
    }

}
/**
 * @param {character[][]} board
 * @param {number[]} click
 * @return {character[][]}
 */
var updateBoard = function (board, click) {
    if (board === null || board.length === 0 || click === null || click.length === 0)
        return board;

    m = board.length;
    n = board[0].length;
    dirs = [[-1, 0], [-1, 1], [0, 1], [1, 1], [1, 0], [1, -1], [0, -1], [-1, -1]]; // U, UR, R, DR, D, DL, L, LU

    if (board[click[0]][click[1]] === 'M') {
        board[click[0]][click[1]] = 'X';
        return board;
    }

    dfs(board, click[0], click[1]);

    // let queue = [];
    // queue.push([click[0] , click[1]]);
    // board[click[0]][click[1]] = 'B';
    // while(queue.length > 0){
    //     let curr = queue.shift();
    //     let mineCount = getMines(board, curr[0], curr[1]);

    //     if(mineCount > 0){
    //         board[curr[0]][curr[1]] = mineCount.toString();;
    //     } else {
    //         for(let i=0; i<dirs.length; i++){
    //             let nr = curr[0] + dirs[i][0];
    //             let nc = curr[1] + dirs[i][1]; 
    //             if(nr >= 0 && nc >= 0 && nr < m && nc < n && board[nr][nc] === 'E'){
    //                 board[nr][nc] = "B";
    //                 queue.push([nr,nc]);
    //             }
    //         }
    //     }
    // }

    return board;
};